nondex-common/src/main/java/edu/illinois/nondex/common/SystematicRandom.java

```diff
@@ -0,0 +1,142 @@
+                    lines = Files.readAllLines(Paths.get(logFileName));
+                } catch (IOException ioe) {
+                    ioe.printStackTrace();
+                }
+                Object[] choiceValues = new Object[3];
+                for (String element: lines) {
+                    String delimit = "[ ]+";
+                    if (!element.isEmpty()) {
+                        choiceValues = element.split(delimit);
+                        int last = Integer.parseInt(choiceValues[0].toString());
+                        int max = Integer.parseInt(choiceValues[1].toString());
+                        boolean explore = Boolean.parseBoolean(choiceValues[2].toString());
+                        StackElement lastMax = new StackElement( last, max, explore );
+                        choice.push(lastMax);
+                    }
+                }
+            }
+        } else {
+            System.err.println("The env variable does not exist. File could not be created");
+            System.exit(1);
+        }
+    }
+
+    public int nextInt(int max) {
+        int last;
+        count++;
+        boolean explore;
+        if (replayIndex < choice.size()) {
+            last =  choice.get(replayIndex).getLast();
+        } else {
+            last = 0;
+            if (count > STARTING_COUNT + 1) {
+                explore = true;
+            } else {
+                explore = false;
+            }
+            StackElement choiceNums = new StackElement(last, max, explore);
+            choice.push(choiceNums);
+        }
+        replayIndex++;
+        return last;
+    }
+
+    public void endRun() throws IOException {
+        while (!choice.isEmpty()) {
+            ArrayList<Object> ex = new ArrayList<>();
+            StackElement lastMax = choice.pop();
+            int last = lastMax.getLast();
+            int max = lastMax.getMax();
+            boolean explore = false;
+            if (last < max - 1) {
+                last++;
+                if (choice.size() > STARTING_COUNT) {
+                    explore = true;
+                }
+                StackElement lm = new StackElement(last, max, explore);
+                choice.push(lm);
+                replayIndex = 0;
+                if (Files.exists(Paths.get(logFileName))) {
+                    Files.delete(Paths.get(logFileName));
+                }
+                BufferedWriter bufferedWriter = Files.newBufferedWriter(Paths.get(logFileName));
+                for (StackElement element : choice) {
+                    String lastAndMax = element.getLast() + " " + element.getMax() + " " + element.getExplore();
+                    bufferedWriter.write(lastAndMax);
+                    bufferedWriter.newLine();
+                }
+                bufferedWriter.close();
+                return;
+            }
+            for (StackElement ch: choice) {
+                ex.add(ch.getExplore());
+            }
+            if (!ex.contains(true)) {
```

Comment:
You seem to be using this ex only as a flag - if it was ever set, you go into this if. Why not just use a boolean flag, for example on line 134 you could do: flag = flag || ch.getExplore();